---
title: "tpm_calc_combinedSex"
author: "Melanie Smith"
date: "20 March 2024"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    message = FALSE,
    warning = FALSE,
    cache = FALSE,
    fig.align = "center",
    results = "hide"
)

```

This script converts the counts generated by `featureCounts` into transcripts per million (TPMs).  
Conversion to TPMs is required before we can compare the bulk counts with scRNAseq counts.  
In this case, the comparison will be used for cell type deconvolution.  

```{r}
library(IRanges)
library(GenomicFeatures)
library(edgeR)
# library(tidyverse)
library(readxl)
library(readr)
library(magrittr)

# set project directory
projectDir <- "/home/smit1924/20231213_GDM_smit1924"
expt_name <- "melanie_v2_0_3"

#genome_gtf_infile <- file.path(projectDir, "clearBoxCleanData/gencode.v29.annotation.gtf")
#genome_gtf_infile <- file.path(projectDir, "rawData/gencode.v32.chr_patch_hapl_scaff.annotation.gtf") # I originally used a gtf file but gff3 is also possible
#gencodev29_gff_inFile <- file.path(projectDir, "clearBoxCleanData/gencode.v29.chr_patch_hapl_scaff.annotation.gff3")
gencodev32_gff_inFile <- file.path(projectDir, "rawData/gencode.v32.chr_patch_hapl_scaff.annotation.gff3")
input_counts_female <- file.path(projectDir, "rawData/20231206_GDM_female_readCounts.txt")
input_counts_male <- file.path(projectDir, "rawData/20231206_GDM_male_readCounts.txt")
input_file_metadata <- file.path(projectDir, "cleanData/GDM_metadata_cleaned.csv")

outdir <- file.path(projectDir, "cleanData")
# output_female_cibersort_file <- file.path(outdir, paste0(expt_name, "_bulkCountMatrix_female.txt"))
# output_male_cibersort_file <- file.path(outdir, paste0(expt_name, "_bulkCountMatrix_male.txt"))
output_all_cibersort_file <- file.path(outdir, "tpm_bulkCountMatrix_all.txt")
output_all_cibersort_file_gene_symbol <- file.path(outdir, "tpm_bulkCountMatrix_all_gene_symbol.txt")

# dir.create(outdir)

# set max digits
options(digits=3)

# set filtering criteria
filterCPM <- 2
numSamples <- 5
```

Before I can transform my raw counts to TPMs I need to calculate the gene lengths from the original .gtf file.   

```{r}

# Try it
# Specifics of calculating rpkm: https://support.bioconductor.org/p/64585/

# 1. Get gene lengths
# https://www.biostars.org/p/83901/
# Or using GTFtools, but keep chromosome limitations in mind:
# http://genomespot.blogspot.com/2019/01/using-gtf-tools-to-get-gene-lengths.html
# GTFtools offers flexibility about how to calculate the gene length (median, mean, etc.)

# Note, this is the gtf used in the RNA-Seq pipeline to annotatate hg38
txdb <- GenomicFeatures::makeTxDbFromGFF(file = file.path(gencodev32_gff_inFile), format = "gff3")
# Collect exons per gene id
exon_list_per_gene <- GenomicFeatures::exonsBy(txdb, by = "gene")
exon_list_per_gene


```

```{r}

# For each gene, reduce all exons to a set of non-overlapping exons, calculate their length (widths) and sum them
exonic_gene_sizes <- 
  sum(IRanges::width(IRanges::reduce(exon_list_per_gene))) %>% 
  tibble::as_tibble(rownames = NA) %>% 
  tibble::rownames_to_column(var = "gene") %>%
  dplyr::select(., gene, length = value)
exonic_gene_sizes

```

## Import metadata

```{r}
## Metadata
# import metadata
metadata <- read_csv(file = input_file_metadata, col_names = TRUE) %>%
  # create the samplename column
  mutate(
    samplename = str_pad(
      case_when(
        Cohort == "STOP" ~ paste0("STP", str_pad(ID, width = 4, pad = "0")),
        Cohort == "SCOPE" ~ paste0("SCP", str_pad(ID, width = 4, pad = "0")),
        TRUE ~ NA_character_
      ),
      width = 7,
      side = "right",
      pad = "0"
    )
  ) %>%
  select(samplename, everything())
male <- dplyr::filter(metadata, FetalSex == "Male") %>%
  dplyr::select(., samplename)
female <- dplyr::filter(metadata, FetalSex == "Female") %>%
  dplyr::select(., samplename)
```

## Import count files
### Import -s 2 counts (female)

```{r import gene counts female}

# import the counts table
# counts are annotated to GRCh38
rawCounts_female <- read.delim(file = file.path(input_file_counts_female)) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_female) <- gsub("X.scratch.user.smit1924.20231206_GDM_female_grch38.aligned_data.|_PlacRNA_|
                                 _GRCh38_Aligned.sortedByCoord.out.bam|_S.*|_",
                                 "",
                                 colnames(rawCounts_female))
# remove the genes from the PAR_Y (all have zero counts)
rawCounts_female %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
# remove the number after the period
#rawCounts_female %<>% tidyr::separate(., col = Geneid, into = c("ensembl", "right")) #%>%
 # dplyr::select(., -right) %>%
 # tibble::column_to_rownames("ensembl")

dim(rawCounts_female)

```

### Import -s 2 counts (male)

```{r import gene counts male}

# import the counts table
# counts are annotated to GRCh38
rawCounts_male <- read.delim(file = file.path(input_file_counts_male)) %>%
  as.data.frame()
# tidy up the column names
colnames(rawCounts_male) <- gsub("X.scratch.user.smit1924.20231206_GDM_male_grch38.aligned_data.|_PlacRNA_|
                                 _GRCh38_Aligned.sortedByCoord.out.bam|_S.*|_",
                                 "",
                                 colnames(rawCounts_male))
# remove the genes from the PAR_Y (all have zero counts)
rawCounts_male %<>% dplyr::filter(., !grepl("PAR_Y",Geneid))
# remove the number after the period
# rawCounts_male %<>% tidyr::separate(., col = Geneid, into = c("ensembl", "right")) %>%
#   dplyr::select(., -right) %>%
#   tibble::column_to_rownames("ensembl")

dim(rawCounts_male)

```

## Combine counts data into a single df

```{r combine counts, echo=FALSE}

# test to make sure the rownames are the same
identical(rownames(rawCounts_male), rownames(rawCounts_female))

# combine the counts tables
rawCounts_combinedSex <- dplyr::left_join(rawCounts_male,
                                             rawCounts_female,
                                             by = "Geneid")
dim(rawCounts_combinedSex)

rownames(rawCounts_combinedSex) <- NULL # Resets numbers to start at one after omitting first four rows loading data
colnames(rawCounts_combinedSex)[1] <- "gene_id" # rename the gene column to match Nick's code

```

```{r}

# 2. Process raw count data in edgeR again, as per melanieRawCountsProcessing_v2.0.3.rmd
# Add gene length to expression data
raw_data <- rawCounts_combinedSex %>%
  dplyr::mutate(Length = exonic_gene_sizes$length[match(gene_id, exonic_gene_sizes$gene)])

raw_data <- raw_data %>% 
  tibble::column_to_rownames("gene_id")
head(raw_data)
dim(raw_data) # 58676, 62

y <- edgeR::DGEList(counts = raw_data %>% dplyr::select(-Length),
             genes = raw_data %>% dplyr::select(Length))

y
y$genes # A Length column is present 
dim(y)

head(y$counts) # Gene names are set as rownames
y$samples$lib.size


# Identify genes expressed at > 1 cpm in at least one sample
keep <- rowSums(cpm(y) > filterCPM) >= numSamples
table(keep) # Keeping 13767 genes

# Apply filter
y <- y[keep, , keep.lib.sizes = FALSE]
dim(y) # 13767    61
y$samples$lib.size
# TMM normalisation 
y$samples$norm.factors # currently all normalisation factors == 1
y <- calcNormFactors(y, method = "TMM")
y$samples$norm.factors # 1.098 0.922 1.029 1.006 0.854 1.054 1.037 1.072 1.114 0.913 1.121 0.994 1.032 0.905 0.967 1.022 1.034 1.090 1.024 0.944 0.902 0.978 0.926 1.020 1.156 1.003 1.051 1.027 0.995 0.964 1.045 0.951 0.949 1.089 1.193 1.000 1.020 1.058 0.974 0.895 0.941 1.036 1.098 0.963 0.895 1.004 0.931 0.945 1.163 0.954 0.969 1.010 0.955 0.924 0.937 1.051 0.945 0.987 1.067 0.975 0.974

# Now calculate RPKM: https://support.bioconductor.org/p/64585/
rpkm <- rpkm(y)

# And TPM: https://www.biostars.org/p/388584/#394579
tpm <- t( t(rpkm) / colSums(rpkm) ) * 1e6
# These TPM values now incorporate the TMM factors

```

## lets add the gene name information

### Import gencode V32 gff3 file

```{r}

# import the gencodev32_gff_inFile file
all_gencode_v32 <- rtracklayer::import(gencodev32_gff_inFile)
# this file contains more information than we need here
# subset out only the columns we need
gene_data <- data.frame(ensembl_gene_id = all_gencode_v32@elementMetadata$gene_id,
                        hgnc_symbol = all_gencode_v32@elementMetadata$gene_name,
                        seqnames = all_gencode_v32@seqnames)
# we're left with multiple identical rows, keep one each
gencode_v32_gene_id_symbol_chr_biotype <- gene_data %>%
  dplyr::distinct(, .keep_all = TRUE)

```

### add the gene info to the tpms

```{r}
# join the counts with the gene info
tpm_gene_id_symbol_chr_biotype <- data.frame(tpm) %>%
  tibble::rownames_to_column("ensembl_gene_id") %>%
  dplyr::left_join(., gencode_v32_gene_id_symbol_chr_biotype, by = "ensembl_gene_id") %>%
  dplyr::select(colnames(gencode_v32_gene_id_symbol_chr_biotype), colnames(data.frame(tpm)))

# save a copy of the tpm counts with the gene info
tpm_gene_id_symbol_chr_biotype %>%
  # remove the number after the period
  tidyr::separate(., col = ensembl_gene_id, into = c("gene", "right")) %>% 
  dplyr::select(., -right) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cibersort_file,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

# save a copy of the tpm counts with only the hgnc gene symbol
tpm_gene_id_symbol_chr_biotype %>%
  # remove the ensembl gene id and the chromosome name
  dplyr::select(., -ensembl_gene_id, -seqnames) %>%
  # save TPM count file for CIBERSTORTx
  write.table(file = output_all_cibersort_file_gene_symbol,
              quote = FALSE,
              sep = "\t",
              col.names = TRUE,
              row.names = FALSE)

```

## Session information

```{r session info}
sessionInfo()
```